Core/src/Geometry/CuboidVolumeStack.cpp

```diff
@@ -0,0 +1,893 @@
+    return mid(direction) - halfLength(direction);
+  }
+  double max(AxisDirection direction) const {
+    return mid(direction) + halfLength(direction);
+  }
+
+  void set(
+      std::initializer_list<std::pair<CuboidVolumeBounds::BoundValues, double>>
+          keyValues) {
+    updatedBounds->set(keyValues);
+  }
+
+  void setLocalTransform(const Transform3& transform,
+                         const Transform3& groupTransform) {
+    localTransform = transform;
+    globalTransform = groupTransform * localTransform;
+    transformDirty = true;
+  }
+
+  void commit(const Logger& logger) {
+    // make a copy so we can't accidentally modify in-place
+    auto copy = std::make_shared<CuboidVolumeBounds>(*updatedBounds);
+
+    std::optional<Transform3> transform = std::nullopt;
+    if (transformDirty) {
+      transform = globalTransform;
+    }
+
+    volume->update(std::move(updatedBounds), transform, logger);
+    bounds = copy.get();
+    updatedBounds = std::move(copy);
+    transformDirty = false;
+  }
+};
+
+CuboidVolumeStack::CuboidVolumeStack(std::vector<Volume*>& volumes,
+                                     AxisDirection direction,
+                                     AttachmentStrategy strategy,
+                                     ResizeStrategy resizeStrategy,
+                                     const Logger& logger)
+    : Volume(initialVolume(volumes)),
+      m_dir(direction),
+      m_resizeStrategy(resizeStrategy),
+      m_volumes(volumes) {
+  auto dirIdx = static_cast<std::size_t>(m_dir);
+  m_dirOrth1 = static_cast<AxisDirection>((dirIdx + 1) % 3);
+  m_dirOrth2 = static_cast<AxisDirection>((dirIdx + 2) % 3);
+
+  initializeOuterVolume(strategy, logger);
+}
+
+CuboidVolumeStack::CuboidVolumeStack(std::vector<Volume*>& volumes,
+                                     Vector3 direction,
+                                     AttachmentStrategy strategy,
+                                     ResizeStrategy resizeStrategy,
+                                     const Logger& logger)
+    : Volume(initialVolume(volumes)),
+      m_resizeStrategy(resizeStrategy),
+      m_volumes(volumes) {
+  // TODO: What's a good tolerance here?
+  const auto& tolerance = s_onSurfaceTolerance;
+  Vector3 localDirVector =
+      m_volumes.front()->transform().rotation().inverse() * direction;
+  if ((localDirVector - Vector3::UnitX()).norm() < tolerance) {
+    m_dir = AxisDirection::AxisX;
+  } else if ((localDirVector - Vector3::UnitY()).norm() < tolerance) {
+    m_dir = AxisDirection::AxisY;
+  } else if ((localDirVector - Vector3::UnitZ()).norm() < tolerance) {
+    m_dir = AxisDirection::AxisZ;
+  } else {
+    throw std::invalid_argument("CuboidVolumeStack: Invalid axis direction");
+  }
+
+  auto dirIdx = static_cast<std::size_t>(m_dir);
+  m_dirOrth1 = static_cast<AxisDirection>((dirIdx + 1) % 3);
+  m_dirOrth2 = static_cast<AxisDirection>((dirIdx + 2) % 3);
+
+  initializeOuterVolume(strategy, logger);
```

Comment:
You could put the determination of the direction from the `Vector3` into a (static) function, and call the other constructor directly from this one, something like:

```cpp
CuboidVolumeStack::CuboidVolumeStack(std::vector<Volume*>& volumes,
                                     Vector3 direction,
                                     AttachmentStrategy strategy,
                                     ResizeStrategy resizeStrategy,
                                     const Logger& logger)
    : CuboidVolumeStack(volumes, 
                        directionFromVector(direction), 
                        strategy, resizeStrategy, logger) {}
```